File: arrow/sam/retrieve.cc
#include <iostream>

#include <arrow/array.h>
#include <arrow/io/memory.h>
#include <arrow/ipc/reader.h>
#include <arrow/record_batch.h>
#include <arrow/util/logging.h>
#include <plasma/client.h>

std::shared_ptr<arrow::RecordBatch> GetRecordBatchFromPlasma(plasma::ObjectID object_id,
                                                             plasma::PlasmaClient &client)
{
    plasma::ObjectBuffer object_buffer;
    ARROW_CHECK_OK(client.Get(&object_id, 1, -1, &object_buffer));

    auto buffer = object_buffer.data;

    arrow::io::BufferReader buffer_reader(buffer);
    std::shared_ptr<arrow::ipc::RecordBatchReader> record_batch_stream_reader;
    ARROW_CHECK_OK(arrow::ipc::RecordBatchStreamReader::Open(&buffer_reader, &record_batch_stream_reader));

    std::shared_ptr<arrow::RecordBatch> record_batch;
    arrow::Status status = record_batch_stream_reader->ReadNext(&record_batch);

    return record_batch;
}

int main(int argc, char **argv)
{
    plasma::ObjectID object_id = plasma::ObjectID::from_binary("0FF1CE00C0FFEE00BEEF");

    // Start up and connect a Plasma client
    plasma::PlasmaClient client;
    ARROW_CHECK_OK(client.Connect("/tmp/store"));

    std::shared_ptr<arrow::RecordBatch> record_batch = GetRecordBatchFromPlasma(object_id, client);

    // Disconnect the Plasma client
    ARROW_CHECK_OK(client.Disconnect());

    std::cout << "record_batch->column(0)->ToString(): "
              << record_batch->column(0)->ToString() << std::endl;
}

File: arrow/sam/retrieve.py
import numpy as np
import pandas as pd
import pyarrow as pa
import pyarrow.plasma as plasma
import binascii

# https://arrow.apache.org/docs/python/plasma.html#getting-pandas-dataframes-from-plasma


def get_record_batch_from_plasma(object_id, client):
    [buffer] = client.get_buffers([object_id])
    buffer_reader = pa.BufferReader(buffer)
    reader = pa.RecordBatchStreamReader(buffer_reader)
    record_batch = reader.read_next_batch()

    return record_batch


if __name__ == "__main__":
    hex_strs = ['ac2f75c043fbc36709d315f2245746d8588c3ac1',
                '25eb8c48ff89cb854fc09081cc47edfc8619b214',
                'a80fed48162bd24b6807a2b15f4bd52f3f1fda94']

    bin_strs = [binascii.unhexlify(hex_str) for hex_str in hex_strs]

    object_ids = [plasma.ObjectID(bin_str) for bin_str in bin_strs]

    client = plasma.connect("/tmp/plasma")

    record_batches = [get_record_batch_from_plasma(
        object_id, client) for object_id in object_ids]

    dfs = [record_batch.to_pandas() for record_batch in record_batches]
    all_records = pd.concat(dfs)
    all_records.sort_values(by=['group_name', 'seq'], inplace=True)
